File: step_13_binary_trees/13.2_medium/P001_height_of_binary_tree/Solution.java
/**
 * Problem: Height of Binary Tree / Maximum Depth (LeetCode #104)
 * Difficulty: EASY | XP: 10
 *
 * @author DSA_Nova
 */
import java.util.*;

public class Solution {

    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    // ============================================================
    // APPROACH 1 & 3: RECURSIVE DFS (Postorder)
    // Time: O(n)  |  Space: O(h)
    // ============================================================
    public static int maxDepthRecursive(TreeNode root) {
        if (root == null) return 0;
        int leftDepth = maxDepthRecursive(root.left);
        int rightDepth = maxDepthRecursive(root.right);
        return 1 + Math.max(leftDepth, rightDepth);
    }

    // ============================================================
    // APPROACH 2: ITERATIVE BFS (Level-Order)
    // Time: O(n)  |  Space: O(w)
    // ============================================================
    public static int maxDepthBFS(TreeNode root) {
        if (root == null) return 0;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int depth = 0;

        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                if (node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);
            }
            depth++;
        }
        return depth;
    }

    // ============================================================
    // BONUS: Iterative DFS with explicit stack
    // Time: O(n)  |  Space: O(h)
    // ============================================================
    public static int maxDepthIterativeDFS(TreeNode root) {
        if (root == null) return 0;

        Deque<Object[]> stack = new ArrayDeque<>();
        stack.push(new Object[]{root, 1});
        int maxDepth = 0;

        while (!stack.isEmpty()) {
            Object[] pair = stack.pop();
            TreeNode node = (TreeNode) pair[0];
            int depth = (int) pair[1];
            maxDepth = Math.max(maxDepth, depth);

            if (node.right != null) stack.push(new Object[]{node.right, depth + 1});
            if (node.left != null) stack.push(new Object[]{node.left, depth + 1});
        }
        return maxDepth;
    }

    // Helper: build tree from level-order array
    public static TreeNode buildTree(Integer[] arr) {
        if (arr == null || arr.length == 0 || arr[0] == null) return null;
        TreeNode root = new TreeNode(arr[0]);
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int i = 1;
        while (!queue.isEmpty() && i < arr.length) {
            TreeNode curr = queue.poll();
            if (i < arr.length && arr[i] != null) {
                curr.left = new TreeNode(arr[i]);
                queue.offer(curr.left);
            }
            i++;
            if (i < arr.length && arr[i] != null) {
                curr.right = new TreeNode(arr[i]);
                queue.offer(curr.right);
            }
            i++;
        }
        return root;
    }

    public static void main(String[] args) {
        System.out.println("=== Height of Binary Tree ===\n");

        Object[][] tests = {
            {new Integer[]{3, 9, 20, null, null, 15, 7}, 3},
            {new Integer[]{1, null, 2}, 2},
            {new Integer[]{}, 0},
            {new Integer[]{1}, 1},
            {new Integer[]{1, 2, 3, 4, null, null, null, 5}, 4},
        };

        for (Object[] test : tests) {
            Integer[] arr = (Integer[]) test[0];
            int expected = (int) test[1];
            TreeNode root = buildTree(arr);

            int rec = maxDepthRecursive(root);
            int bfs = maxDepthBFS(root);
            int dfs = maxDepthIterativeDFS(root);

            String status = (rec == expected && bfs == expected && dfs == expected) ? "PASS" : "FAIL";
            System.out.printf("Tree: %s%n", Arrays.toString(arr));
            System.out.printf("  Recursive: %d, BFS: %d, Iterative DFS: %d (expected %d) %s%n%n",
                    rec, bfs, dfs, expected, status);
        }
    }
}

File: step_13_binary_trees/13.2_medium/P001_height_of_binary_tree/solution.py
"""
Problem: Height of Binary Tree / Maximum Depth (LeetCode #104)
Difficulty: EASY | XP: 10
"""
from typing import List, Optional
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# ============================================================
# APPROACH 1 & 3: RECURSIVE DFS (Postorder)
# Time: O(n)  |  Space: O(h)
# ============================================================
def max_depth_recursive(root: Optional[TreeNode]) -> int:
    if root is None:
        return 0
    left_depth = max_depth_recursive(root.left)
    right_depth = max_depth_recursive(root.right)
    return 1 + max(left_depth, right_depth)


# ============================================================
# APPROACH 2: ITERATIVE BFS (Level-Order)
# Time: O(n)  |  Space: O(w)
# ============================================================
def max_depth_bfs(root: Optional[TreeNode]) -> int:
    if root is None:
        return 0

    queue = deque([root])
    depth = 0

    while queue:
        for _ in range(len(queue)):
            node = queue.popleft()
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        depth += 1

    return depth


# ============================================================
# BONUS: Iterative DFS with explicit stack
# Time: O(n)  |  Space: O(h)
# ============================================================
def max_depth_iterative_dfs(root: Optional[TreeNode]) -> int:
    if root is None:
        return 0

    stack = [(root, 1)]
    max_d = 0

    while stack:
        node, depth = stack.pop()
        max_d = max(max_d, depth)
        if node.right:
            stack.append((node.right, depth + 1))
        if node.left:
            stack.append((node.left, depth + 1))

    return max_d


def build_tree(arr: List[Optional[int]]) -> Optional[TreeNode]:
    if not arr or arr[0] is None:
        return None
    root = TreeNode(arr[0])
    queue = deque([root])
    i = 1
    while queue and i < len(arr):
        curr = queue.popleft()
        if i < len(arr) and arr[i] is not None:
            curr.left = TreeNode(arr[i])
            queue.append(curr.left)
        i += 1
        if i < len(arr) and arr[i] is not None:
            curr.right = TreeNode(arr[i])
            queue.append(curr.right)
        i += 1
    return root


if __name__ == "__main__":
    print("=== Height of Binary Tree ===\n")

    tests = [
        ([3, 9, 20, None, None, 15, 7], 3),
        ([1, None, 2], 2),
        ([], 0),
        ([1], 1),
        ([1, 2, 3, 4, None, None, None, 5], 4),
    ]

    for arr, expected in tests:
        root = build_tree(arr)
        rec = max_depth_recursive(root)
        bfs = max_depth_bfs(root)
        dfs = max_depth_iterative_dfs(root)
        status = "PASS" if rec == expected == bfs == dfs else "FAIL"
        print(f"Tree: {arr}")
        print(f"  Recursive: {rec}, BFS: {bfs}, Iterative DFS: {dfs} (expected {expected}) {status}\n")
